### ISSUE TITLE
JSPM: JSDom dependency causing error?

### ISSUE THREAD

**Issue by mrbannon** [paperjs/paper.js#1104]
Hello all,

I've just started trying to use paper@0.10.2 (from previously using paper@0.9.25). After installing (jspm install npm:paper@0.10.2) and including paper in my code, I'm getting the following console error (in Chrome):

`GET http://132.206.14.136:9001/@empty/lib/jsdom/living/generated/utils.js 404 (Not Found)`

Any help is greatly appreciated.

Tx,

Ryan

**Comment by lehni** [comment]
What's jspm? Are you running paper.js in your browser? There's no need for JSDOM in the browser. Webpack handles this correctly.

**Comment by mrbannon** [comment]
JavaScript package manager. It installs node packages for use in a browser.

Yes, I'm using paper in a browser (Version 51.0.2704.103 (64-bit) on Mac).

**Comment by lehni** [comment]
We haven't tested with JSPM so far. Is there a way to tell JSPM to not load a module? JSDom isn't required in the browser, it's required on NodeJS to emulate the browser environment, and Webpack respects these settings:

```json
  "browser": {
    "canvas": false,
    "jsdom": false,
    "source-map-support": false,
    "./dist/node/window.js": false,
    "./dist/node/extend.js": false
  },
```

 I'm getting tired of JS package management. There are way too many ways of doing these things out there...

**Comment by mrbannon** [comment]
No argument here. :)

Tx for the response. I'll keep working on it.

**Comment by lehni** [comment]
I'm happy to try and find a solution. The problem is a complex one, since we're now offering one "hybrid" version of paper.js that is expected to work in Node.js, Electron, Browsers and Web Workers.  Could you let me know how you're installing JSPM and getting a simple test case up and running? I could then look into what can be done. Maybe it's easy to fix (fingers crossed).

**Comment by mrbannon** [comment]
My previous comment had an error in it. Has been corrected.

**Comment by lehni** [comment]
Thanks for the info. I still would like to try and replicate this locally, but I'mnew to JSPM. How do set things up to get to setup that I can test?

**Comment by mrbannon** [comment]
I'll try to create a minimum project to recreate the error.

**Comment by lehni** [comment]
Using the bower endpoint may be another solution:
https://www.npmjs.com/package/jspm-bower-endpoint

**Comment by mrbannon** [comment]
The less new things I have to learn, the better ;).

BTW, thanks for 0.10. view.translate works like a charm!

**Comment by mrbannon** [comment]
@lehni Please take a look at https://github.com/mrbannon/jspm_npm_test. The README has instructions on how to recreate the issue. Any questions please message me. :)

**Comment by lehni** [comment]
Hmmm... I've tired to fix it now, by adding this line: https://github.com/paperjs/paper.js/commit/80f4ba947f881973d81c109d7796ca75dfcbeefd#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R80

Bit it looks like JSPM (or browserify?) isn't ver smart about how it handles packages that shouldn't be loaded. Due to `"jsdom": false` it replaced `jsdom` in `"jsdom/lib/jsdom/living/generated/utils"` with `@empty`, and then appears to ingore the separate statement `"jsdom/lib/jsdom/living/generated/utils": false`

**Comment by mrbannon** [comment]
I'll give it a try. Stay tuned...

**Comment by lehni** [comment]
I think JSPM needs a change, to understand that "jsdom/lib/jsdom/living/generated/utils" is part of the "jsdom" dependency and hence shouldn't be loaded, and replaced fully with "@empty".

**Comment by mrbannon** [comment]
I would agree.

**Comment by mrbannon** [comment]
Is your fix already in the npm package?

**Comment by lehni** [comment]
No, the fix isn't working. I tested it with a temporary branch... Not pushing a new release.

And I think I know why `"paper": "github:paperjs/paper.js@0.10.2"` works for you:

I'm currently not including the `dist/node/*` sub-modules in the git releases (tags). These only get published to NPM (from `src/node/*`), and added to `prebuilt/module` and `prebuilt/dist`. So if you take the module from a git tag, it won't have these modules ready and they'll probably be `@empty` then?

https://github.com/paperjs/paper.js/commit/a02f181c00ae59451a75e8b7345e727da8f8382d

**Comment by mrbannon** [comment]
That makes sense.

**Comment by lehni** [comment]
Hmm but it's weird.We also have `"./dist/node/window.js": false,` and that still gets loaded...

**Comment by mrbannon** [comment]
This is why I get nervous when installing new dependencies ;)

**Comment by lehni** [comment]
I opened this now: https://github.com/jspm/jspm-cli/issues/1981

Let's see how it goes.

**Comment by lehni** [comment]
@mrbannon good news! this will be fixed in jspm 0.17: https://github.com/jspm/jspm-cli/issues/1981#issuecomment-234297532

And actually, my temporary fix should also work, I just didn't test it properly... Can you live with this until our next release?

**Comment by mrbannon** [comment]
Oh yeah, for sure! That's good news. Thanks for the hard work!

Ryan

**Comment by lehni** [comment]
You're welcome!

**Comment by lehni** [comment]
I think I spoke too quickly, reopening again...

**Comment by lehni** [comment]
Please keep the test-case alive for now.

**Comment by lehni** [comment]
Thanks to https://github.com/jspm/jspm-cli/issues/1981#issuecomment-234513446, JSPM now works with the already published v0.10.2!
